pandas FutureWarning: Setting an item of incompatible dtype is deprecated

With pandas 2.1.0:
```
    petab.lint_problem(petab_problem)
/home/runner/.local/lib/python3.9/site-packages/petab/lint.py:888: in lint_problem
    errors_occurred |= validate_visualization_df(problem)
/home/runner/.local/lib/python3.9/site-packages/petab/visualize/lint.py:39: in validate_visualization_df
    _apply_defaults(vis_df)
/home/runner/.local/lib/python3.9/site-packages/petab/visualize/lint.py:131: in _apply_defaults
    set_default(C.PLOT_TYPE_DATA, C.MEAN_AND_SD)
/home/runner/.local/lib/python3.9/site-packages/petab/visualize/lint.py:127: in set_default
    vis_df[column].fillna(value, inplace=True)
/home/runner/.local/lib/python3.9/site-packages/pandas/core/generic.py:7210: in fillna
    new_data = self._mgr.fillna(
/home/runner/.local/lib/python3.9/site-packages/pandas/core/internals/base.py:173: in fillna
    return self.apply_with_block(
/home/runner/.local/lib/python3.9/site-packages/pandas/core/internals/managers.py:354: in apply
    applied = getattr(b, f)(**kwargs)
/home/runner/.local/lib/python3.9/site-packages/pandas/core/internals/blocks.py:1415: in fillna
    nbs = self.putmask(mask.T, value, using_cow=using_cow)
/home/runner/.local/lib/python3.9/site-packages/pandas/core/internals/blocks.py:1232: in putmask
    return self.coerce_to_target_dtype(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = NumpyBlock: 7 dtype: float64, other = 'MeanAndSD', warn_on_upcast = True

    @final
    def coerce_to_target_dtype(self, other, warn_on_upcast: bool = False) -> Block:
        """
        coerce the current block to a dtype compat for other
        we will return a block, possibly object, and not raise
    
        we can also safely try to coerce to the same dtype
        and will receive the same block
        """
        new_dtype = find_result_type(self.values.dtype, other)
    
        # In a future version of pandas, the default will be that
        # setting `nan` into an integer series won't raise.
        if (
            is_scalar(other)
            and is_integer_dtype(self.values.dtype)
            and isna(other)
            and other is not NaT
        ):
            warn_on_upcast = False
        elif (
            isinstance(other, np.ndarray)
            and other.ndim == 1
            and is_integer_dtype(self.values.dtype)
            and is_float_dtype(other.dtype)
            and lib.has_only_ints_or_nan(other)
        ):
            warn_on_upcast = False
    
        if warn_on_upcast:
>           warnings.warn(
                f"Setting an item of incompatible dtype is deprecated "
                "and will raise in a future error of pandas. "
                f"Value '{other}' has dtype incompatible with {self.values.dtype}, "
                "please explicitly cast to a compatible dtype first.",
                FutureWarning,
                stacklevel=find_stack_level(),
            )
E           FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'MeanAndSD' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
```

--- a/.flake8
+++ b/.flake8
@@ -1,8 +1,12 @@
 [flake8]
 
 extend-ignore =
+    # whitespace before ":", conflicts with black
+    E203
     F403
     F405
+    # line too long, conflicts with black in rare cases
+    E501
 exclude =
     build,
     dist,

--- a/petab/calculate.py
+++ b/petab/calculate.py
@@ -83,7 +83,7 @@ def calculate_residuals_for_table(
     # create residual df as copy of measurement df, change column
     residual_df = measurement_df.copy(deep=True).rename(
         columns={MEASUREMENT: RESIDUAL})
-
+    residual_df[RESIDUAL] = residual_df[RESIDUAL].astype("float64")
     # matching columns
     compared_cols = set(MEASUREMENT_DF_COLS)
     compared_cols -= {MEASUREMENT}

--- a/petab/simplify.py
+++ b/petab/simplify.py
@@ -73,11 +73,11 @@ def condition_parameters_to_parameter_table(problem: Problem):
             continue
 
         series = problem.condition_df[parameter_id]
-        value = petab.to_float_if_float(series[0])
+        value = petab.to_float_if_float(series.iloc[0])
 
         # same value for all conditions and no parametric overrides (str)?
         if isinstance(value, float) and len(series.unique()) == 1:
-            replacements[parameter_id] = series[0]
+            replacements[parameter_id] = series.iloc[0]
 
     if not replacements:
         return

--- a/petab/visualize/data_overview.py
+++ b/petab/visualize/data_overview.py
@@ -58,6 +58,8 @@ def get_data_per_observable(measurement_df: pd.DataFrame) -> pd.DataFrame:
     """
 
     my_measurements = measurement_df.copy()
+    my_measurements[PREEQUILIBRATION_CONDITION_ID] = my_measurements[PREEQUILIBRATION_CONDITION_ID].astype("object")
+
     index = [SIMULATION_CONDITION_ID]
     if PREEQUILIBRATION_CONDITION_ID in my_measurements:
         my_measurements[PREEQUILIBRATION_CONDITION_ID].fillna('', inplace=True)

--- a/petab/visualize/lint.py
+++ b/petab/visualize/lint.py
@@ -124,6 +124,8 @@ def _apply_defaults(vis_df: pd.DataFrame):
         if column not in vis_df:
             vis_df[column] = value
         elif value is not None:
+            if isinstance(value, str):
+                vis_df[column] = vis_df[column].astype('object')
             vis_df[column].fillna(value, inplace=True)
 
     set_default(C.PLOT_NAME, "")

--- a/tests/test_conditions.py
+++ b/tests/test_conditions.py
@@ -26,6 +26,7 @@ def test_get_parametric_overrides():
 
     assert conditions.get_parametric_overrides(condition_df) == []
 
+    condition_df['fixedParameter1'] = condition_df['fixedParameter1'].astype("object")
     condition_df.loc[0, 'fixedParameter1'] = 'parameterId'
 
     assert conditions.get_parametric_overrides(condition_df) == ['parameterId']

--- a/tests/test_lint.py
+++ b/tests/test_lint.py
@@ -469,7 +469,7 @@ def test_check_parameter_df():
         LOWER_BOUND: [1e-5, 1e-6, 1e-7],
         UPPER_BOUND: [1e5, 1e6, 1e7]
     }).set_index(PARAMETER_ID)
-
+    parameter_df[NOMINAL_VALUE] = parameter_df[NOMINAL_VALUE].astype("object")
     lint.check_parameter_df(df=parameter_df)
 
     # NOMINAL_VALUE empty, for non-estimated parameter

--- a/tests/test_petab.py
+++ b/tests/test_petab.py
@@ -24,6 +24,7 @@ def condition_df_2_conditions():
         'fixedParameter1': [1.0, 2.0]
     })
     condition_df.set_index('conditionId', inplace=True)
+    condition_df.fixedParameter1 = condition_df.fixedParameter1.astype("object")
     return condition_df
 
 

